src/web/hooks/use-widget-state.ts

```diff
@@ -2,6 +2,48 @@ import { useCallback, useEffect, useState, type SetStateAction } from "react";
 import type { UnknownObject } from "../types.js";
 import { useOpenAiGlobal } from "./use-openai-global.js";
 
+const WIDGET_CONTEXT_KEY = "__widget_context" as const;
+
+function filterWidgetContext<T extends UnknownObject>(
+  state?: T | null
+): T | null {
+  if (state === null) {
+    return null;
+  }
+
+  const { [WIDGET_CONTEXT_KEY]: _, ...filteredState } = state as T & {
+    [WIDGET_CONTEXT_KEY]?: unknown;
+  };
+
+  return filteredState as T;
+}
+
+function persistWidgetContext<T extends UnknownObject>(
```

Comment:
sorry, I removed my comment beforehand, you were right: "preserve" was better than "persist" here :)